core/src/main/java/com/datastrato/gravitino/catalog/CapabilityHelpers.java

```diff
@@ -25,22 +33,165 @@ public static TableChange[] applyCapabilities(Capability capabilities, TableChan
+    return Arrays.stream(transforms)
+        .map(t -> applyCapabilities(t, capabilities))
+        .toArray(Transform[]::new);
+  }
+
+  public static Distribution applyCapabilities(Distribution distribution, Capability capabilities) {
+    applyCapabilities(distribution.expressions(), capabilities);
+    return distribution;
+  }
+
+  public static SortOrder[] applyCapabilities(SortOrder[] sortOrders, Capability capabilities) {
+    return Arrays.stream(sortOrders)
+        .map(s -> applyCapabilities(s, capabilities))
+        .toArray(SortOrder[]::new);
+  }
+
+  public static Index[] applyCapabilities(Index[] indexes, Capability capabilities) {
+    return Arrays.stream(indexes)
+        .map(i -> applyCapabilities(i, capabilities))
+        .toArray(Index[]::new);
+  }
+
+  public static Namespace applyCapabilities(
+      Namespace namespace, Capability.Scope identScope, Capability capabilities) {
+    String metalake = namespace.level(0);
+    String catalog = namespace.level(1);
+    if (identScope == Capability.Scope.TABLE
+        || identScope == Capability.Scope.FILESET
+        || identScope == Capability.Scope.TOPIC) {
+      String schema = namespace.level(namespace.length() - 1);
+      schema = applyCaseSensitiveOnName(Capability.Scope.SCHEMA, schema, capabilities);
+      applyNameSpecification(Capability.Scope.SCHEMA, schema, capabilities);
+      return Namespace.of(metalake, catalog, schema);
+    }
+    return namespace;
+  }
+
+  private static Index applyCapabilities(Index index, Capability capabilities) {
+    String[][] fieldNames = index.fieldNames();
+    for (String[] fieldName : fieldNames) {
+      String[] sensitiveOnColumnName = applyCaseSensitiveOnColumnName(fieldName, capabilities);
+      fieldName[0] = sensitiveOnColumnName[0];
+      applyNameSpecification(Capability.Scope.COLUMN, sensitiveOnColumnName[0], capabilities);
+    }
+    return index;
+  }
+
+  private static Transform applyCapabilities(Transform transform, Capability capabilities) {
+    for (Expression arg : transform.arguments()) {
+      applyCapabilities(arg, capabilities);
+    }
+    return transform;
+  }
+
+  private static SortOrder applyCapabilities(SortOrder sortOrder, Capability capabilities) {
+    applyCapabilities(sortOrder.expression(), capabilities);
+    return sortOrder;
+  }
+
+  private static Expression[] applyCapabilities(Expression[] expressions, Capability capabilities) {
+    return Arrays.stream(expressions)
+        .map(e -> applyCapabilities(e, capabilities))
+        .toArray(Expression[]::new);
+  }
+
+  private static Expression applyCapabilities(Expression expression, Capability capabilities) {
+    if (expression instanceof NamedReference.FieldReference) {
+      NamedReference.FieldReference ref = (NamedReference.FieldReference) expression;
+      String[] fieldName = applyCaseSensitiveOnColumnName(ref.fieldName(), capabilities);
+      ref.fieldName()[0] = fieldName[0];
```

Comment:
Also here, I think it is not good to do in place modification, do you have better solution?

Reply:
refactored